File: 10_10_2024_Maximum_Width_Ramp.java
class Solution {
    public int maxWidthRamp(int[] nums) {
        int n = nums.length;
        Stack<Integer> s = new Stack<>();
        
        for (int i = 0; i < n; ++i) {
            if (s.isEmpty() || nums[s.peek()] > nums[i]) {
                s.push(i);
            }
        }
        
        int maxWidth = 0;
        
        for (int j = n - 1; j >= 0; --j) {
            while (!s.isEmpty() && nums[s.peek()] <= nums[j]) {
                maxWidth = Math.max(maxWidth, j - s.pop());
            }
        }
        
        return maxWidth;
    }
}

File: 10_10_2024_Maximum_Width_Ramp.py
class Solution:
    def maxWidthRamp(self, nums):
        n = len(nums)
        s = []
        
        for i in range(n):
            if not s or nums[s[-1]] > nums[i]:
                s.append(i)
        
        maxWidth = 0
        
        for j in range(n - 1, -1, -1):
            while s and nums[s[-1]] <= nums[j]:
                maxWidth = max(maxWidth, j - s.pop())
        
        return maxWidth

